homeassistant/helpers/script.py

```diff
@@ -297,22 +297,34 @@ async def _async_event_step(self):
             self._action[CONF_EVENT], event_data, context=self._context
         )
 
-    async def _async_condition_step(self):
-        """Test if condition is matching."""
-        config_cache_key = frozenset((k, str(v)) for k, v in self._action.items())
+    async def _check_condition(self, cond):
+        config_cache_key = frozenset((k, str(v)) for k, v in cond.items())
         config = self._config_cache.get(config_cache_key)
         if not config:
-            config = await condition.async_from_config(self._hass, self._action, False)
+            config = await condition.async_from_config(self._hass, cond, False)
             self._config_cache[config_cache_key] = config
 
+        return config(self._hass, self._variables)
+
+    async def _async_condition_step(self):
+        """Test if condition is matching."""
         self._script.last_action = self._action.get(
             CONF_ALIAS, self._action[CONF_CONDITION]
         )
-        check = config(self._hass, self._variables)
+        check = await self._check_condition(self._action)
         self._log("Test condition %s: %s", self._script.last_action, check)
         if not check:
             raise _StopScript
 
+    async def _async_branch_step(self):
+        branches = self._action.get("branch", [])
+        for b in branches:
+            condition = b.get("if")
+            if not condition or await self._check_condition(condition):
```

Comment:
Would the await here prevent the condition checks from being atomic? i.e. the conditions for different cases would be evaluated in a potentially different context.

Reply:
the await make that the function will be executed. Otherwise the code will do nothing. With means that will be stop this execution and come in some time back if the mainloop will that. We are async at anytime.

Reply:
Yes, my point is not that we should remove the await, but rather that if we want to ensure atomicity, we'll need to work around this somehow. I'm referring [to this comment from the arch topic](https://github.com/home-assistant/architecture/issues/378#issuecomment-623891643):
> As long as there is no async await in the code path it will be frozen.

Reply:
Yes, and with our core design, you can't run the steps without await. So you can not archive that without complete rewrite how the core work.

Reply:
You're right, but we can try to figure out another approach to ensure atomicity in this case.

Reply:
Our core don't support that, which basically need, you need write an external application like AppDaemon get the the hole state machine and relay just with this state table while the full program run. This approach is fine for 95% of the use case, simple but effective.

Reply:
I don't know this area, well enough, but we could, for example, copy all of the states and have `_check_condition` except a copy of all states. It's definitely possible. Not sure how easy it is, but it's doable, and is just one example.

Reply:
Good point.
The `break` below will protect from executing multiple branches, but it's quite possible that a state change between two condition evaluations cases no branch to be executed.

It seems the `_check_conditions` asynchronously produces synchronous checkers and then run them. That should be possible to split into two parts...

Reply:
This is possible. If you look at the code, the `await` is only used to gather the conditions. So you can do 1 await cycle to get all condition functions and then execute them all without await.

Reply:
Exactly what I had in mind.